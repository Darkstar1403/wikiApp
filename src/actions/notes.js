import { addDoc, collection, updateDoc, doc, deleteDoc } from "firebase/firestore";
import Swal from "sweetalert2";
import { db } from "../components/firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes, loadNote } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () =>{
    return async(dispatch, getState) =>{
        try{
            const {uid} = getState().auth;
            const newNote = {
                title: '',
                body: '',
                date: new Date().getTime()
            }
            
            const docRef = await addDoc(collection(db, `${uid}/journal/notes`),newNote);
            dispatch(activeNote(docRef.id, newNote));
            dispatch(addNewNotes(docRef.id, newNote));
        }
        catch(error){
            Swal.fire('Error', 'Registrar nuevo articulo error', 'error');
        }

    }
}

export const loadActiveNote = (id) =>{
    return async(dispatch) =>{
        try {
            const note = await loadNote(id);
            dispatch(activeNote(note));
            console.log(note);
        } catch (error) {
            Swal.fire('Error', 'Error al cargar el articulo', 'error');
        }
    }
}

export const activeNote = (note) => ({
    type: types.articleActive,
    payload: {
        ...note
    }
});

export const unsetActiveNote = () => ({type: types.articleUnsetActive});

export const addNewNotes = (id, note) =>({
    type:types.articleAddNew,
    payload:{
        id,
        ...note
    }
})

export const startLoadingNotes = () =>{
    return async(dispatch) =>{
        try{
            const notes = await loadNotes();
            console.log(notes);
            dispatch(setNotes(notes));
        }
        catch(error){
            Swal.fire('Error', 'La carga de los articulos fallÃ³', 'error');
        }
    }
}

export const setNotes = (notes) =>({
    type: types.articleLoad,
    payload: notes
});

export const startSaveNote = (note) =>{
    return async() =>{
        try{
            if(!note.img){
                delete note.img;
            }
            const noteToFirestore = {...note};
            //asi es como eliminas una llave de un objeto
            delete noteToFirestore.id;
            const docRef = await addDoc(collection(db, `/articles`),note);
            console.log(docRef);
            Swal.fire('Articulo agregado', note.title, 'success')
        }
        catch(error){
            Swal.fire('Error', 'Error al guardar el articulo', 'error');
        }

    }
}

export const startUpdateNote = (note) =>{
    return async(dispatch) =>{
        try{
            if(!note.img){
                delete note.img;
            }
            const noteToFirestore = {...note};
            //asi es como eliminas una llave de un objeto
            delete noteToFirestore.id;
            const refDoc = doc(db, `/articles/${note.id}`);
            await updateDoc(refDoc, noteToFirestore);
            Swal.fire('Se ha modificado', note.title, 'success');
            dispatch(activeNote(note));
        }
        catch(error){
            Swal.fire('Error', 'Saving note failed', 'error');
        }

    }
}

export const refreshNote = (id, note) =>({
    type: types.articleUpdated,
    payload : {
        id, 
        note: {
            id,
            ...note
        }
    }
});
//todos los actions start se refieren a tareas asincronas
export const startUploading = (file) =>{
    return async(dispatch, getState) =>{
        const {active} = getState().article;
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () =>{
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        active.img = fileUrl;

        dispatch(activeNote(active))

        Swal.close();
    }
}

export const startDeleting = (id) =>{
    return async(dispatch, getState) =>{
        try{
            const {uid} = getState().auth;
            const refDoc = doc(db, `/${uid}/journal/notes/${id}`);
            await deleteDoc(refDoc);
            dispatch(deleteNote(id));
            Swal.fire('Deleted', 'Eliminado correctamente', 'success')
        }
        catch(error){
            Swal.fire('Error', 'Error al eliminar el articulo', 'error');
        }

    }
}

export const deleteNote = (id) => ({
    type: types.articleDelete,
    payload: id
});

export const noteLogout = () =>({
    type:types.articleLogout
});