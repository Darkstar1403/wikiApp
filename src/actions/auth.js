import Swal from 'sweetalert2';
import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../components/firebase/firebase-config';
import {types} from '../types/types';
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword = (email, password) =>{
    return (dispatch) =>{
        dispatch(startLoading());
        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
        .then(({user})=>{
            if(user.photoURL){
                console.log(user.photoURL);
            }
            (user.photoURL)?dispatch(login(user.uid, user.displayName, true)):dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
            Swal.fire(
                "Has ingresado a la mejor wiki",
                'Presionaste el botón',
                'success'
              );
        }).catch(e =>{
            console.log(e);
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        });
        
    }
}

export const  startRegister = (email, password, name, admin) =>{
    return (dispatch) =>{
        dispatch(startLoading());
        const auth = getAuth();
        let url = null;
        if(admin){
            url = 'admin'
        }
        createUserWithEmailAndPassword(auth, email, password).then(async ({user})=>{
            await updateProfile(user, {displayName:name, photoURL: url});
            dispatch(login(user.uid, user.displayName, (url)?true:false));
            dispatch(finishLoading());
            Swal.fire(
                "Te has registrado en la mejor wiki",
                'Presionaste el botón',
                'success'
              );
        }).catch(e =>{
            console.log(e);
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        })
        
    }
}

export const startGoogleLogin = () =>{
    return (dispatch) =>{
        dispatch(startLoading());
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then(({user}) =>{
                dispatch(login(user.uid, user.displayName))
            }).catch(e => console.log(e));
            dispatch(finishLoading());
        
    }
}

export const login = (uid, displayName, admin = false) =>(
    {
        type:types.login,
        payload: {
            uid,
            displayName,
            admin
        }
    }
)

export const startLogout = () =>{
    return async(dispatch) =>{
        const auth = getAuth();
       await signOut(auth);
       dispatch(logout());
       dispatch(noteLogout());
    }
}

export const logout = () =>({
    type:types.logout
})