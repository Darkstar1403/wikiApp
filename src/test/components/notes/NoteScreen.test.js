import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import '@testing-library/jest-dom'
import { activeNote } from "../../../actions/notes";
import { NoteScreen } from "../../../components/notes/NoteScreen";


jest.mock('../../../actions/notes', () =>({
    startNewNote:jest.fn(),
    activeNote:jest.fn()
}))
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth:{
        uid: '1',
        displayName: 'Branco'
    },
    ui:{
        loading: false,
        msgError: null
    },
    notes:{
        active:{
            id:1234,
            title:'Hola',
            body:'Mundo',
            date:1234
        },
        notes:[]

    }
};
let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Tests of <NoteScreem/>', () => {
    const wrapper = mount(
        <Provider store={store}>
                <NoteScreen/>
        </Provider>
    )
    test('should be render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    });

    test('should use activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target:{
                name:'title',
                value:'HIiiii'
            }
        });
        expect(activeNote).toHaveBeenLastCalledWith(1234,
            {
                body: 'Mundo',
                title:'HIiiii',
                id:1234,
                date:1234
            });
    })
    
})
