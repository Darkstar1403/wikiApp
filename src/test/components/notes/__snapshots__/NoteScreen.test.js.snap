// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of <NoteScreem/> should be render correctly 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      September 8th 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        <i
          class="fas fa-camera fa-lg"
        />
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content animate__animated animate__bounceIn"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value="Mundo"
    />
    <div
      className="notes__image"
    >
      <img />
    </div>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
